cpp/src/community/leiden_impl.cuh

```diff
@@ -42,17 +45,508 @@ void check_clustering(graph_view_t<vertex_t, edge_t, false, multi_gpu> const& gr
+                                                         cluster_keys.begin(),
+                                                         cluster_weights.begin());
+
+      louvain_of_refined_partition.resize(0, handle.get_stream());
+      louvain_of_refined_partition.shrink_to_fit(handle.get_stream());
+
+      tmp_weights_buffer.resize(0, handle.get_stream());
+      tmp_weights_buffer.shrink_to_fit(handle.get_stream());
+
+      cluster_keys.resize(
+        static_cast<size_t>(thrust::distance(cluster_keys.begin(), pair_of_iterators_end.first)),
+        handle.get_stream());
+      cluster_weights.resize(static_cast<size_t>(thrust::distance(cluster_weights.begin(),
+                                                                  pair_of_iterators_end.second)),
+                             handle.get_stream());
+
+      cluster_keys.shrink_to_fit(handle.get_stream());
+      cluster_weights.shrink_to_fit(handle.get_stream());
+
+      if constexpr (graph_view_t::is_multi_gpu) {
+        rmm::device_uvector<vertex_t> tmp_keys_buffer(0, handle.get_stream());  // #C
+
+        std::tie(tmp_keys_buffer, tmp_weights_buffer) =
+          shuffle_ext_vertex_value_pairs_to_local_gpu_by_vertex_partitioning(
+            handle, std::move(cluster_keys), std::move(cluster_weights));
+
+        thrust::sort_by_key(handle.get_thrust_policy(),
+                            tmp_keys_buffer.begin(),
+                            tmp_keys_buffer.end(),
+                            tmp_weights_buffer.begin());
+
+        cluster_keys.resize(tmp_keys_buffer.size(), handle.get_stream());
+        cluster_weights.resize(tmp_weights_buffer.size(), handle.get_stream());
+
+        pair_of_iterators_end = thrust::reduce_by_key(handle.get_thrust_policy(),
+                                                      tmp_keys_buffer.begin(),
+                                                      tmp_keys_buffer.end(),
+                                                      tmp_weights_buffer.begin(),
+                                                      cluster_keys.begin(),
+                                                      cluster_weights.begin());
+
+        tmp_keys_buffer.resize(0, handle.get_stream());
+        tmp_keys_buffer.shrink_to_fit(handle.get_stream());
+        tmp_weights_buffer.resize(0, handle.get_stream());
+        tmp_weights_buffer.shrink_to_fit(handle.get_stream());
+
+        cluster_keys.resize(
+          static_cast<size_t>(thrust::distance(cluster_keys.begin(), pair_of_iterators_end.first)),
+          handle.get_stream());
+        cluster_weights.resize(static_cast<size_t>(thrust::distance(cluster_weights.begin(),
+                                                                    pair_of_iterators_end.second)),
+                               handle.get_stream());
+
+        cluster_keys.shrink_to_fit(handle.get_stream());
+        cluster_weights.shrink_to_fit(handle.get_stream());
```

Comment:
This code becomes unnecessary if you count # uniques first and resize cluster_keys and cluster_weights properly.